File: LeetCode/Patterns/Binary Search/1056. Capacity To Ship Packages Within D Days/Solution.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int left=0;
        int right=0;
        for(int i=0;i<weights.length;i++){
            left=Math.max(left,weights[i]);
            right=right+weights[i];
        }
        while(left<=right){
            int mid=left+(right-left)/2;
            int noOfDays=1;
            int currLoad=0;
            for(int i=0;i<weights.length;i++){
                if(currLoad+weights[i]<=mid){
                    currLoad+=weights[i];
                }else{
                    noOfDays++;
                    currLoad=weights[i];
                }
            }
            if(noOfDays<=days){
                right=mid-1;
            }else{
                left=mid+1;
            }
        }
        return left;
        
    }
}

File: LeetCode/Patterns/Binary Search/1056. Capacity To Ship Packages Within D Days/Solution.py
class Solution(object):
    def shipWithinDays(self, weights, days):
        left, right = max(weights), sum(weights)
    
        while left < right:
            mid = (left + right) // 2
            if canShip(weights, days, mid):
                right = mid
            else:
                left = mid + 1
            
        return left

def canShip(weights, days, capacity):
    total = 0
    required_days = 1
    for w in weights:
        if total + w > capacity:
            required_days += 1
            total = 0
        total += w
    return required_days <= days
        
        
        